core/services/offchainreporting/data_source.go

```diff
@@ -22,14 +24,43 @@ type dataSource struct {
 
 var _ ocrtypes.DataSource = (*dataSource)(nil)
 
-// The context passed in here has a timeout of observationTimeout.
+// The context passed in here has a timeout of (ObservationTimeout - ObservationGracePeriod).
+// Upon context cancellation, its expected that we return any usable values within ObservationGracePeriod.
 func (ds dataSource) Observe(ctx context.Context) (ocrtypes.Observation, error) {
-	results, err := ds.pipelineRunner.ExecuteAndInsertNewRun(ctx, ds.spec, ds.ocrLogger)
+	var observation ocrtypes.Observation
+	start := time.Now()
+	run, err := pipeline.NewRun(ds.spec, start)
 	if err != nil {
-		return nil, errors.Wrapf(err, "error executing new run for job ID %v", ds.jobID)
+		return observation, errors.Wrapf(err, "error creating new run for spec ID %v", ds.spec.ID)
 	}
 
-	result, err := results.SingularResult()
+	trrs, err := ds.pipelineRunner.ExecuteRun(ctx, run, ds.ocrLogger)
+	if err != nil {
+		return observation, errors.Wrapf(err, "error executing run for spec ID %v", ds.spec.ID)
+	}
+	end := time.Now()
+
+	run.FinishedAt = &end
+
+	finalResult := trrs.FinalResult()
+	run.Outputs = finalResult.OutputsDB()
+	run.Errors = finalResult.ErrorsDB()
+
+	// Do the database write in a non-blocking fashion
+	// so we can return the observation results immediately.
+	// This is helpful in the case of a blocking API call, where
+	// we reach the passed in context deadline and we want to
+	// immediately return any result we have and do not want to have
+	// a db write block that.
+	go func() {
```

Comment:
I don't know whether you care, but note that:
- This could leak go-routines if one of the underlying functions does not terminate
- Perform inserts out of order

Reply:
Might be good to chat with spook or sam about this as well, I recall them having opinions on this in the past.

Reply:
Yeah for 1) the timeout is DatabaseMaximumTxDuration inside InsertFinishedRunWithResults and 2) insert order doesn't matter for these pipeline run results. I did talk to Sam about this before doing it but cc @samsondav @spooktheducks do you guys have any further thoughts on it?

Reply:
Insert order doesn't matter much as long as we do our `SELECT ... ORDER BY` queries correctly elsewhere in the app which we should be doing anyway.

I am not too worried about leaking goroutines as long as have some kind of context timeout on the DB insert then there is an upper bound.

Reply:
More ideal would be a queue with backpressure but that's probably overengineering it at this point.